File: js/src/red_hat_boy_states.ts
import { IPoint, Sound, Audio } from "./engine";
import { LOOPING } from "./sound";

const FLOOR: number = 449;
const HEIGHT: number = 570;
const PLAYER_HEIGHT: number = HEIGHT - FLOOR;
const STARTING_POINT: number = -20;

const IDLE_FRAME_NAME: string = "Idle";
const RUN_FRAME_NAME: string = "Run";
const SLIDING_FRAME_NAME: string = "Slide";
const JUMPING_FRAME_NAME: string = "Jump";
const FALLING_FRAME_NAME: string = "Dead";

const IDLE_FRAMES: number = 29;
const RUNNING_FRAMES: number = 23;
export const SLIDING_FRAMES: number = 14;
export const JUMPING_FRAMES: number = 35;
const FALLING_FRAMES: number = 29;

const RUNNING_SPEED: number = 4;
const JUMP_SPEED: number = -25;

const GRAVITY: number = 1;
const TERMINAL_VELOCITY: number = 20;

export interface IRedHatBoyContext {
  frame: number;
  position: IPoint;
  velocity: IPoint;
  audio: Audio;
  jump_sound: Sound;
}

export class RedHatBoyContext implements IRedHatBoyContext {
  constructor(
    public frame: number,
    public position: IPoint,
    public velocity: IPoint,
    public audio: Audio,
    public jump_sound: Sound
  ) {}

  update(frame_count: number): RedHatBoyContext {
    if (this.frame < frame_count) {
      this.frame += 1;
    } else {
      this.frame = 0;
    }
    if (this.velocity.y < TERMINAL_VELOCITY) {
      this.velocity.y += GRAVITY;
    }
    this.position.y += this.velocity.y;
    if (this.position.y > FLOOR) {
      this.position.y = FLOOR;
    }
    return this;
  }

  reset_frame(): RedHatBoyContext {
    this.frame = 0;
    return this;
  }

  run_right(): RedHatBoyContext {
    this.velocity.x += RUNNING_SPEED;
    return this;
  }

  set_vertical_velocity(y: number): RedHatBoyContext {
    this.velocity.y = y;
    return this;
  }

  stop(): RedHatBoyContext {
    this.velocity.x = 0;
    return this;
  }

  set_on(position: number): RedHatBoyContext {
    position = position - PLAYER_HEIGHT;
    this.position.y = position;
    return this;
  }

  play_jump_sound(): RedHatBoyContext {
    this.audio.play_sound(this.jump_sound, LOOPING.NO);
    return this;
  }
}

export enum StateType {
  Idle,
  Running,
  Sliding,
  Jumping,
  Falling,
  KnockedOut,
}

export interface IRedHatBoyState {
  context: IRedHatBoyContext;
  type: StateType;
}

export class IdleRedHatBoyState implements IRedHatBoyState {
  context: RedHatBoyContext;
  type = StateType.Idle;

  constructor(audio: Audio, jump_sound: Sound) {
    this.context = new RedHatBoyContext(0, { x: STARTING_POINT, y: FLOOR }, { x: 0, y: 0 }, audio, jump_sound);
  }

  run(): RunningRedHatBoyState {
    return new RunningRedHatBoyState(this.context.reset_frame().run_right());
  }

  frame_name(): string {
    return IDLE_FRAME_NAME;
  }

  update(): IdleRedHatBoyState {
    this.context = this.context.update(IDLE_FRAMES);
    return this;
  }
}

export class RunningRedHatBoyState implements IRedHatBoyState {
  context: RedHatBoyContext;
  type = StateType.Running;

  constructor(context: RedHatBoyContext) {
    this.context = context;
  }

  frame_name(): string {
    return RUN_FRAME_NAME;
  }

  update(): RunningRedHatBoyState {
    this.context = this.context.update(RUNNING_FRAMES);
    return this;
  }

  slide(): SlidingRedHatBoyState {
    return new SlidingRedHatBoyState(this.context.reset_frame());
  }

  jump(): JumpingRedHatBoyState {
    return new JumpingRedHatBoyState(this.context.set_vertical_velocity(JUMP_SPEED).reset_frame().play_jump_sound());
  }

  knock_out(): FallingRedHatBoyState {
    return new FallingRedHatBoyState(this.context.reset_frame().stop());
  }

  land_on(position: number): RunningRedHatBoyState {
    return new RunningRedHatBoyState(this.context.set_on(position));
  }
}

export class SlidingRedHatBoyState implements IRedHatBoyState {
  context: RedHatBoyContext;
  type = StateType.Sliding;

  constructor(context: RedHatBoyContext) {
    this.context = context;
  }

  frame_name(): string {
    return SLIDING_FRAME_NAME;
  }

  update(): SlidingRedHatBoyState | RunningRedHatBoyState {
    this.context = this.context.update(SLIDING_FRAMES);
    if (this.context.frame >= SLIDING_FRAMES) {
      return this.stand();
    } else {
      return this;
    }
  }

  stand(): RunningRedHatBoyState {
    return new RunningRedHatBoyState(this.context.reset_frame());
  }

  knock_out(): FallingRedHatBoyState {
    return new FallingRedHatBoyState(this.context.reset_frame().stop());
  }

  land_on(position: number): SlidingRedHatBoyState {
    return new SlidingRedHatBoyState(this.context.set_on(position));
  }
}

export class JumpingRedHatBoyState {
  context: RedHatBoyContext;
  type = StateType.Jumping;

  constructor(context: RedHatBoyContext) {
    this.context = context;
  }

  frame_name(): string {
    return JUMPING_FRAME_NAME;
  }

  update(): JumpingRedHatBoyState | RunningRedHatBoyState {
    this.context = this.context.update(JUMPING_FRAMES);
    if (this.context.position.y >= FLOOR) {
      return this.land_on(HEIGHT);
    } else {
      return this;
    }
  }

  land_on(position: number): RunningRedHatBoyState {
    return new RunningRedHatBoyState(this.context.reset_frame().set_on(position));
  }

  knock_out(): FallingRedHatBoyState {
    return new FallingRedHatBoyState(this.context.reset_frame().stop());
  }
}

export class FallingRedHatBoyState implements IRedHatBoyState {
  context: RedHatBoyContext;
  type = StateType.Falling;

  constructor(context: RedHatBoyContext) {
    this.context = context;
  }

  frame_name(): string {
    return FALLING_FRAME_NAME;
  }

  update(): FallingRedHatBoyState | KnockedOutRedHatBoyState {
    this.context = this.context.update(FALLING_FRAMES);
    if (this.context.frame >= FALLING_FRAMES) {
      return this.fall();
    } else {
      return this;
    }
  }

  fall(): KnockedOutRedHatBoyState {
    return new KnockedOutRedHatBoyState(this.context);
  }
}

export class KnockedOutRedHatBoyState implements IRedHatBoyState {
  context: RedHatBoyContext;
  type = StateType.KnockedOut;

  constructor(context: RedHatBoyContext) {
    this.context = context;
  }

  frame_name(): string {
    return FALLING_FRAME_NAME;
  }

  update(): KnockedOutRedHatBoyState {
    return this;
  }
}

export type RedHatBoyState =
  | IdleRedHatBoyState
  | RunningRedHatBoyState
  | SlidingRedHatBoyState
  | JumpingRedHatBoyState
  | FallingRedHatBoyState
  | KnockedOutRedHatBoyState;

File: rust/src/red_hat_boy_states.rs
use crate::engine::Point;

const FLOOR: i16 = 449;
const HEIGHT: i16 = 570;
const PLAYER_HEIGHT: i16 = HEIGHT - FLOOR;
const STARTING_POINT: i16 = -20;

const IDLE_FRAME_NAME: &str = "Idle";
const RUN_FRAME_NAME: &str = "Run";
const SLIDING_FRAME_NAME: &str = "Slide";
const JUMPING_FRAME_NAME: &str = "Jump";
const FALLING_FRAME_NAME: &str = "Dead";

const IDLE_FRAMES: u8 = 29;
const RUNNING_FRAMES: u8 = 23;
pub const SLIDING_FRAMES: u8 = 14;
pub const JUMPING_FRAMES: u8 = 35;
pub const FALLING_FRAMES: u8 = 29;

const RUNNING_SPEED: i16 = 4;
const JUMP_SPEED: i16 = -25;

const GRAVITY: i16 = 1;
const TERMINAL_VELOCITY: i16 = 20;

#[derive(Copy, Clone)]
pub struct RedHatBoyState<S> {
    context: RedHatBoyContext,
    _state: S,
}

#[derive(Copy, Clone)]
pub struct RedHatBoyContext {
    pub frame: u8,
    pub position: Point,
    pub velocity: Point,
}

impl RedHatBoyContext {
    pub fn update(mut self, frame_count: u8) -> Self {
        if self.frame < frame_count {
            self.frame += 1;
        } else {
            self.frame = 0;
        }
        if self.velocity.y < TERMINAL_VELOCITY {
            self.velocity.y += GRAVITY;
        }
        self.position.y += self.velocity.y;
        if self.position.y > FLOOR {
            self.position.y = FLOOR;
        }
        self
    }

    fn reset_frame(mut self) -> Self {
        self.frame = 0;
        self
    }

    fn run_right(mut self) -> Self {
        self.velocity.x += RUNNING_SPEED;
        self
    }

    fn set_vertical_velocity(mut self, y: i16) -> Self {
        self.velocity.y = y;
        self
    }

    fn stop(mut self) -> Self {
        self.velocity.x = 0;
        self
    }

    fn set_on(mut self, position: i16) -> Self {
        let position = position - PLAYER_HEIGHT;
        self.position.y = position;
        self
    }
}

#[derive(Copy, Clone)]
pub struct Idle;
#[derive(Copy, Clone)]
pub struct Running;

#[derive(Copy, Clone)]
pub struct Sliding;

#[derive(Copy, Clone)]
pub struct Jumping;

#[derive(Copy, Clone)]
pub struct Falling;

#[derive(Copy, Clone)]
pub struct KnockedOut;

impl<S> RedHatBoyState<S> {
    pub fn context(&self) -> &RedHatBoyContext {
        &self.context
    }
}

impl RedHatBoyState<Idle> {
    pub fn run(self) -> RedHatBoyState<Running> {
        RedHatBoyState {
            context: self.context.reset_frame().run_right(),
            _state: Running {},
        }
    }
    pub fn new() -> Self {
        RedHatBoyState {
            context: RedHatBoyContext {
                frame: 0,
                position: Point {
                    x: STARTING_POINT,
                    y: FLOOR,
                },
                velocity: Point { x: 0, y: 0 },
            },
            _state: Idle {},
        }
    }

    pub fn frame_name(&self) -> &str {
        IDLE_FRAME_NAME
    }

    pub fn update(mut self) -> Self {
        self.context = self.context.update(IDLE_FRAMES);
        self
    }
}

impl RedHatBoyState<Running> {
    pub fn frame_name(&self) -> &str {
        RUN_FRAME_NAME
    }

    pub fn update(mut self) -> Self {
        self.context = self.context.update(RUNNING_FRAMES);
        self
    }

    pub fn slide(self) -> RedHatBoyState<Sliding> {
        RedHatBoyState {
            context: self.context.reset_frame(),
            _state: Sliding {},
        }
    }

    pub fn jump(self) -> RedHatBoyState<Jumping> {
        RedHatBoyState {
            context: self.context.set_vertical_velocity(JUMP_SPEED).reset_frame(),
            _state: Jumping {},
        }
    }

    pub fn knock_out(self) -> RedHatBoyState<Falling> {
        RedHatBoyState {
            context: self.context.reset_frame().stop(),
            _state: Falling {},
        }
    }

    pub fn land_on(self, position: i16) -> RedHatBoyState<Running> {
        RedHatBoyState {
            context: self.context.set_on(position as i16),
            _state: Running,
        }
    }
}

impl RedHatBoyState<Sliding> {
    pub fn frame_name(&self) -> &str {
        SLIDING_FRAME_NAME
    }

    pub fn update(mut self) -> SlidingEndState {
        self.context = self.context.update(SLIDING_FRAMES);
        if self.context.frame >= SLIDING_FRAMES {
            SlidingEndState::Complete(self.stand())
        } else {
            SlidingEndState::Sliding(self)
        }
    }

    pub fn stand(self) -> RedHatBoyState<Running> {
        RedHatBoyState {
            context: self.context.reset_frame(),
            _state: Running,
        }
    }

    pub fn knock_out(self) -> RedHatBoyState<Falling> {
        RedHatBoyState {
            context: self.context.reset_frame().stop(),
            _state: Falling {},
        }
    }

    pub fn land_on(self, position: i16) -> RedHatBoyState<Sliding> {
        RedHatBoyState {
            context: self.context.set_on(position),
            _state: Sliding,
        }
    }
}

impl RedHatBoyState<Jumping> {
    pub fn frame_name(&self) -> &str {
        JUMPING_FRAME_NAME
    }

    pub fn update(mut self) -> JumpingEndState {
        self.context = self.context.update(JUMPING_FRAMES);
        if self.context.position.y >= FLOOR {
            JumpingEndState::Complete(self.land_on(HEIGHT.into()))
        } else {
            JumpingEndState::Jumping(self)
        }
    }

    pub fn land_on(self, position: i16) -> RedHatBoyState<Running> {
        RedHatBoyState {
            context: self.context.reset_frame().set_on(position),
            _state: Running,
        }
    }

    pub fn knock_out(self) -> RedHatBoyState<Falling> {
        RedHatBoyState {
            context: self.context.reset_frame().stop(),
            _state: Falling {},
        }
    }
}

impl RedHatBoyState<Falling> {
    pub fn frame_name(&self) -> &str {
        FALLING_FRAME_NAME
    }

    pub fn update(mut self) -> FallingEndState {
        self.context = self.context.update(FALLING_FRAMES);
        if self.context.frame >= FALLING_FRAMES {
            FallingEndState::Complete(self.fall())
        } else {
            FallingEndState::Falling(self)
        }
    }

    pub fn fall(self) -> RedHatBoyState<KnockedOut> {
        RedHatBoyState {
            context: self.context,
            _state: KnockedOut {},
        }
    }
}

impl RedHatBoyState<KnockedOut> {
    pub fn frame_name(&self) -> &str {
        FALLING_FRAME_NAME
    }
}

pub enum SlidingEndState {
    Complete(RedHatBoyState<Running>),
    Sliding(RedHatBoyState<Sliding>),
}

pub enum JumpingEndState {
    Complete(RedHatBoyState<Running>),
    Jumping(RedHatBoyState<Jumping>),
}

pub enum FallingEndState {
    Complete(RedHatBoyState<KnockedOut>),
    Falling(RedHatBoyState<Falling>),
}
